File: Emulator/src/main/content/region/fremennik/rellekka/plugin/RockCrabNPC.java
package content.region.fremennik.rellekka.plugin;

import core.game.node.entity.Entity;
import core.game.node.entity.combat.BattleState;
import core.game.node.entity.npc.AbstractNPC;
import core.game.node.entity.npc.agg.AggressiveBehavior;
import core.game.node.entity.npc.agg.AggressiveHandler;
import core.game.world.map.Location;
import core.plugin.Initializable;
import core.tools.RandomFunction;

/**
 * The type Rock crab npc.
 */
@Initializable
public final class RockCrabNPC extends AbstractNPC {

    private static final AggressiveBehavior AGGRO_BEHAVIOR = new AggressiveBehavior() {
        @Override
        public boolean ignoreCombatLevelDifference() {
            return true;
        }

        @Override
        public boolean canSelectTarget(Entity entity, Entity target) {
            return super.canSelectTarget(entity, target) &&
                entity.getLocation().withinDistance(target.getLocation(), 3);
        }
    };

    @Override
    public void onAttack(Entity target) {
        this.aggressor = true;
        this.target = target;
        if (getId() == getOriginalId()) {
            this.transform(getOriginalId() - 1);
        }
    }

    private boolean aggressor;

    private Entity target;

    /**
     * Instantiates a new Rock crab npc.
     *
     * @param id       the id
     * @param location the location
     */
    public RockCrabNPC(int id, Location location) {
        super(id, location, false);
        super.setAggressiveHandler(new AggressiveHandler(this, AGGRO_BEHAVIOR));
        this.setAggressive(true);
        this.setWalks(false);
    }

    /**
     * Instantiates a new Rock crab npc.
     */
    public RockCrabNPC() {
        super(-1, null);
    }

    @Override
    public void handleTickActions() {
        super.handleTickActions();
        if ((aggressor && !inCombat() && target.getLocation().getDistance(this.getLocation()) > 12) || isInvisible()) {
            reTransform();
            aggressor = false;
            target = null;
            getWalkingQueue().reset();
            setWalks(false);
        }
    }

    @Override
    public void sendImpact(BattleState state) {
        if (state.getEstimatedHit() >= 3 && RandomFunction.random(30) != 5) {
            state.setEstimatedHit(0);
        }
        if (state.getEstimatedHit() == 2 && RandomFunction.random(30) != 5) {
            state.setEstimatedHit(0);
        }
    }

    @Override
    public AbstractNPC construct(int id, Location location, Object... objects) {
        return new RockCrabNPC(id, location);
    }

    private int getTransformId() {
        if (getId() != super.getOriginalId()) {
            return getOriginalId();
        }
        return getId() - 1;
    }

    @Override
    public int[] getIds() {
        return new int[]{1266, 1268, 2453, 2890};
    }

}

File: Server/src/main/content/region/fremennik/rellekka/npc/RockCrabNPC.kt
package content.region.fremennik.rellekka.npc

import core.game.node.entity.Entity
import core.game.node.entity.combat.BattleState
import core.game.node.entity.npc.AbstractNPC
import core.game.node.entity.npc.agg.AggressiveBehavior
import core.game.node.entity.npc.agg.AggressiveHandler
import core.game.node.entity.player.Player
import core.game.world.map.Location
import core.game.world.map.RegionManager
import core.plugin.Initializable
import core.tools.RandomFunction
import shared.consts.NPCs

/**
 * Handles the Rock Crab NPC.
 */
@Initializable
class RockCrabNPC(id: Int = -1, location: Location? = null) : AbstractNPC(id, location, false) {

    private var aggressor = false
    private var target: Player? = null

    init {
        if (id != -1 && location != null) {
            aggressiveHandler = AggressiveHandler(this, AGGRO_BEHAVIOR)
            isAggressive = true
            isWalks = false
        }
    }

    override fun onAttack(target: Entity) {
        if (target is Player) {
            this.aggressor = true
            this.target = target
            if (id == originalId) {
                transform(originalId - 1)
            }
        }
    }

    override fun handleTickActions() {
        super.handleTickActions()

        val nearbyPlayers = RegionManager.getLocalPlayers(this, 3)
            .filterIsInstance<Player>()

        if (nearbyPlayers.isNotEmpty() && !inCombat() && !aggressor) {
            aggressor = true
            transform(originalId - 1)

            val nearest = nearbyPlayers.minByOrNull { it.location.getDistance(location) }
            nearest?.let { target = it }
        }

        val closestDistance = nearbyPlayers.minOfOrNull { it.location.getDistance(location) } ?: Double.MAX_VALUE
        if (aggressor && !inCombat() && (closestDistance > 12 || isInvisible)) {
            reTransform()
            aggressor = false
            target = null
            walkingQueue.reset()
            isWalks = false
        }
    }

    override fun sendImpact(state: BattleState) {
        val hit = state.estimatedHit
        if ((hit >= 3 && RandomFunction.random(30) != 5) || (hit == 2 && RandomFunction.random(30) != 5)) {
            state.estimatedHit = 0
        }
    }

    override fun construct(id: Int, location: Location, vararg objects: Any): AbstractNPC {
        return RockCrabNPC(id, location)
    }

    override fun getIds(): IntArray = intArrayOf(
        NPCs.ROCKS_1266,
        NPCs.ROCKS_1268,
        NPCs.BOULDER_2453,
        NPCs.ROCK_2890
    )

    companion object {
        private val AGGRO_BEHAVIOR = object : AggressiveBehavior() {
            override fun ignoreCombatLevelDifference(): Boolean = true

            override fun canSelectTarget(entity: Entity, target: Entity): Boolean {
                return target is Player &&
                        super.canSelectTarget(entity, target) &&
                        entity.location.withinDistance(target.location, 3)
            }
        }
    }
}
